File: geoffrey-main/scripts/secrets.js
#!/usr/bin/env node

/**
 * Geoffrey Secrets Manager
 *
 * Loads secrets exclusively from 1Password CLI.
 *
 * Usage:
 *   const { getSecret, requireSecret, SECRETS } = require('./secrets.js');
 *   const apiKey = requireSecret('OPENAI_API_KEY');
 *   // Or use pre-defined accessors:
 *   const { domain, apiKey } = SECRETS.freshservice;
 *
 * 1Password Setup:
 *   1. Install: brew install --cask 1password-cli
 *   2. Enable CLI integration in 1Password desktop app settings
 *   3. Create "Geoffrey" vault with items matching VAULT_MAP below
 *   4. See docs/1password-setup.md for detailed setup guide
 */

const { spawnSync } = require('child_process');

// Map environment variable names to 1Password secret references
// Format: op://vault/item/field
const VAULT_MAP = {
  // Freshservice
  FRESHSERVICE_DOMAIN: 'op://Geoffrey/Freshservice/domain',
  FRESHSERVICE_API_KEY: 'op://Geoffrey/Freshservice/api-key',

  // Research LLMs
  OPENAI_API_KEY: 'op://Geoffrey/OpenAI/api-key',
  GEMINI_API_KEY: 'op://Geoffrey/Gemini/api-key',
  PERPLEXITY_API_KEY: 'op://Geoffrey/Perplexity/api-key',
  XAI_API_KEY: 'op://Geoffrey/XAI/api-key',

  // Google Workspace
  GOOGLE_CLIENT_ID: 'op://Geoffrey/Google-Workspace/client-id',
  GOOGLE_CLIENT_SECRET: 'op://Geoffrey/Google-Workspace/client-secret',

  // ElevenLabs
  ELEVENLABS_API_KEY: 'op://Geoffrey/ElevenLabs/api-key',

  // Obsidian MCP
  OBSIDIAN_API_KEY: 'op://Geoffrey/Obsidian-MCP/api-key',

  // Red Rover
  REDROVER_USERNAME: 'op://Geoffrey/RedRover/username',
  REDROVER_PASSWORD: 'op://Geoffrey/RedRover/password',
  REDROVER_API_KEY: 'op://Geoffrey/RedRover/api-key',
};

// Cache for loaded secrets (avoid repeated CLI calls)
const secretsCache = new Map();

// Track 1Password availability check
let _1pChecked = false;
let _1pAvailable = false;

/**
 * Check if 1Password CLI is available and authenticated
 */
function is1PasswordAvailable() {
  if (_1pChecked) return _1pAvailable;

  try {
    const result = spawnSync('op', ['account', 'list'], { encoding: 'utf8' });
    _1pAvailable = result.status === 0;
  } catch {
    _1pAvailable = false;
  }
  _1pChecked = true;
  return _1pAvailable;
}

/**
 * Ensure 1Password is available, throw helpful error if not
 */
function ensure1Password() {
  if (!is1PasswordAvailable()) {
    throw new Error(
      '1Password CLI is not available or not authenticated.\n\n' +
        'Setup required:\n' +
        '  1. Install: brew install --cask 1password-cli\n' +
        '  2. Enable CLI integration in 1Password app:\n' +
        '     Settings → Developer → Enable CLI integration\n' +
        '  3. Authenticate: op signin\n\n' +
        'See docs/1password-setup.md for detailed instructions.'
    );
  }
}

/**
 * Load a secret from 1Password
 */
function loadFrom1Password(secretRef) {
  try {
    const result = spawnSync('op', ['read', secretRef], {
      encoding: 'utf8',
      timeout: 10000,
    });
    if (result.status === 0 && result.stdout) {
      return result.stdout.trim();
    }
    return null;
  } catch (error) {
    return null;
  }
}

/**
 * Get a single secret by name
 *
 * @param {string} name - Environment variable name (e.g., 'OPENAI_API_KEY')
 * @returns {string|null} - Secret value or null if not found
 */
function getSecret(name) {
  // Check cache first
  if (secretsCache.has(name)) {
    return secretsCache.get(name);
  }

  // Ensure 1Password is available
  ensure1Password();

  // Get 1Password reference
  const secretRef = VAULT_MAP[name];
  if (!secretRef) {
    throw new Error(
      `Unknown secret: ${name}\n` +
        `Available secrets: ${Object.keys(VAULT_MAP).join(', ')}`
    );
  }

  // Load from 1Password
  const value = loadFrom1Password(secretRef);

  // Cache the result
  if (value) {
    secretsCache.set(name, value);
  }

  return value;
}

/**
 * Load multiple secrets at once
 *
 * @param {string[]} names - Array of environment variable names
 * @returns {Object} - Object with secret names as keys
 */
function loadSecrets(names) {
  const result = {};
  for (const name of names) {
    result[name] = getSecret(name);
  }
  return result;
}

/**
 * Get all configured secrets (useful for debugging)
 */
function listAvailableSecrets() {
  return Object.keys(VAULT_MAP);
}

/**
 * Require a secret (throws if not found)
 *
 * @param {string} name - Environment variable name
 * @returns {string} - Secret value
 * @throws {Error} - If secret is not found
 */
function requireSecret(name) {
  const value = getSecret(name);
  if (!value) {
    const opRef = VAULT_MAP[name] || 'not configured';
    throw new Error(
      `Missing required secret: ${name}\n` +
        `1Password reference: ${opRef}\n\n` +
        'To add this secret:\n' +
        `  1. Open 1Password\n` +
        `  2. Create/edit item in Geoffrey vault matching: ${opRef}\n` +
        `  3. See docs/1password-setup.md for vault structure`
    );
  }
  return value;
}

// Pre-defined secret accessors for common use cases
const SECRETS = {
  get freshservice() {
    return {
      domain: requireSecret('FRESHSERVICE_DOMAIN'),
      apiKey: requireSecret('FRESHSERVICE_API_KEY'),
    };
  },
  get openai() {
    return requireSecret('OPENAI_API_KEY');
  },
  get gemini() {
    return requireSecret('GEMINI_API_KEY');
  },
  get perplexity() {
    return requireSecret('PERPLEXITY_API_KEY');
  },
  get xai() {
    return requireSecret('XAI_API_KEY');
  },
  get google() {
    return {
      clientId: requireSecret('GOOGLE_CLIENT_ID'),
      clientSecret: requireSecret('GOOGLE_CLIENT_SECRET'),
    };
  },
  get elevenlabs() {
    return requireSecret('ELEVENLABS_API_KEY');
  },
  get obsidian() {
    return requireSecret('OBSIDIAN_API_KEY');
  },
  get redrover() {
    return {
      username: requireSecret('REDROVER_USERNAME'),
      password: requireSecret('REDROVER_PASSWORD'),
      apiKey: requireSecret('REDROVER_API_KEY'),
    };
  },
};

module.exports = {
  getSecret,
  loadSecrets,
  requireSecret,
  listAvailableSecrets,
  is1PasswordAvailable,
  ensure1Password,
  SECRETS,
  VAULT_MAP,
};

File: geoffrey-main/scripts/secrets.py
#!/usr/bin/env python3
# /// script
# requires-python = ">=3.11"
# dependencies = []
# ///
"""
Geoffrey Secrets Manager (Python)

Loads secrets exclusively from 1Password CLI.

Usage:
    from secrets import get_secret, require_secret, SECRETS
    api_key = require_secret('OPENAI_API_KEY')
    # Or use pre-defined accessors:
    api_key = SECRETS.openai

1Password Setup:
    1. Install: brew install --cask 1password-cli
    2. Enable CLI integration in 1Password desktop app settings
    3. Create "Geoffrey" vault with items matching VAULT_MAP below
    4. See docs/1password-setup.md for detailed setup guide
"""

import subprocess
import sys
from functools import lru_cache
from typing import Optional

# Map environment variable names to 1Password secret references
# Format: op://vault/item/field
VAULT_MAP = {
    # Freshservice
    "FRESHSERVICE_DOMAIN": "op://Geoffrey/Freshservice/domain",
    "FRESHSERVICE_API_KEY": "op://Geoffrey/Freshservice/api-key",

    # Research LLMs
    "OPENAI_API_KEY": "op://Geoffrey/OpenAI/api-key",
    "GEMINI_API_KEY": "op://Geoffrey/Gemini/api-key",
    "PERPLEXITY_API_KEY": "op://Geoffrey/Perplexity/api-key",
    "XAI_API_KEY": "op://Geoffrey/XAI/api-key",

    # Google Workspace
    "GOOGLE_CLIENT_ID": "op://Geoffrey/Google-Workspace/client-id",
    "GOOGLE_CLIENT_SECRET": "op://Geoffrey/Google-Workspace/client-secret",

    # ElevenLabs
    "ELEVENLABS_API_KEY": "op://Geoffrey/ElevenLabs/api-key",

    # Obsidian MCP
    "OBSIDIAN_API_KEY": "op://Geoffrey/Obsidian-MCP/api-key",
}

# Cache for loaded secrets
_secrets_cache: dict[str, str] = {}


@lru_cache(maxsize=1)
def is_1password_available() -> bool:
    """Check if 1Password CLI is available and authenticated."""
    try:
        subprocess.run(
            ["op", "account", "list"],
            capture_output=True,
            check=True,
            timeout=5,
        )
        return True
    except (subprocess.CalledProcessError, FileNotFoundError, subprocess.TimeoutExpired):
        return False


def ensure_1password() -> None:
    """Ensure 1Password is available, raise helpful error if not."""
    if not is_1password_available():
        raise RuntimeError(
            "1Password CLI is not available or not authenticated.\n\n"
            "Setup required:\n"
            "  1. Install: brew install --cask 1password-cli\n"
            "  2. Enable CLI integration in 1Password app:\n"
            "     Settings → Developer → Enable CLI integration\n"
            "  3. Authenticate: op signin\n\n"
            "See docs/1password-setup.md for detailed instructions."
        )


def _load_from_1password(secret_ref: str) -> Optional[str]:
    """Load a secret from 1Password."""
    try:
        result = subprocess.run(
            ["op", "read", secret_ref],
            capture_output=True,
            text=True,
            check=True,
            timeout=10,
        )
        return result.stdout.strip()
    except (subprocess.CalledProcessError, FileNotFoundError, subprocess.TimeoutExpired):
        return None


def get_secret(name: str) -> Optional[str]:
    """
    Get a single secret by name.

    Args:
        name: Environment variable name (e.g., 'OPENAI_API_KEY')

    Returns:
        Secret value or None if not found
    """
    # Check cache first
    if name in _secrets_cache:
        return _secrets_cache[name]

    # Ensure 1Password is available
    ensure_1password()

    # Get 1Password reference
    secret_ref = VAULT_MAP.get(name)
    if not secret_ref:
        raise ValueError(
            f"Unknown secret: {name}\n"
            f"Available secrets: {', '.join(VAULT_MAP.keys())}"
        )

    # Load from 1Password
    value = _load_from_1password(secret_ref)

    # Cache the result
    if value:
        _secrets_cache[name] = value

    return value


def load_secrets(names: list[str]) -> dict[str, Optional[str]]:
    """
    Load multiple secrets at once.

    Args:
        names: List of environment variable names

    Returns:
        Dict with secret names as keys
    """
    return {name: get_secret(name) for name in names}


def require_secret(name: str) -> str:
    """
    Require a secret (raises if not found).

    Args:
        name: Environment variable name

    Returns:
        Secret value

    Raises:
        ValueError: If secret is not found
    """
    value = get_secret(name)
    if not value:
        op_ref = VAULT_MAP.get(name, "not configured")
        raise ValueError(
            f"Missing required secret: {name}\n"
            f"1Password reference: {op_ref}\n\n"
            f"To add this secret:\n"
            f"  1. Open 1Password\n"
            f"  2. Create/edit item in Geoffrey vault matching: {op_ref}\n"
            f"  3. See docs/1password-setup.md for vault structure"
        )
    return value


def list_available_secrets() -> list[str]:
    """Get all configured secret names."""
    return list(VAULT_MAP.keys())


class SecretsAccessor:
    """Pre-defined secret accessors for common use cases."""

    @property
    def freshservice_domain(self) -> str:
        return require_secret("FRESHSERVICE_DOMAIN")

    @property
    def freshservice_api_key(self) -> str:
        return require_secret("FRESHSERVICE_API_KEY")

    @property
    def openai(self) -> str:
        return require_secret("OPENAI_API_KEY")

    @property
    def gemini(self) -> str:
        return require_secret("GEMINI_API_KEY")

    @property
    def perplexity(self) -> str:
        return require_secret("PERPLEXITY_API_KEY")

    @property
    def xai(self) -> str:
        return require_secret("XAI_API_KEY")

    @property
    def google_client_id(self) -> str:
        return require_secret("GOOGLE_CLIENT_ID")

    @property
    def google_client_secret(self) -> str:
        return require_secret("GOOGLE_CLIENT_SECRET")

    @property
    def elevenlabs(self) -> str:
        return require_secret("ELEVENLABS_API_KEY")

    @property
    def obsidian(self) -> str:
        return require_secret("OBSIDIAN_API_KEY")


# Singleton accessor
SECRETS = SecretsAccessor()


if __name__ == "__main__":
    # CLI for testing
    import argparse

    parser = argparse.ArgumentParser(description="Geoffrey Secrets Manager")
    parser.add_argument("--list", action="store_true", help="List available secrets")
    parser.add_argument("--get", metavar="NAME", help="Get a specific secret")
    parser.add_argument("--check-1p", action="store_true", help="Check 1Password availability")

    args = parser.parse_args()

    if args.list:
        print("Available secrets:")
        for name in list_available_secrets():
            print(f"  {name}: {VAULT_MAP[name]}")
    elif args.get:
        try:
            value = require_secret(args.get)
            # Mask middle of secret for security
            if len(value) > 8:
                masked = value[:4] + "..." + value[-4:]
            else:
                masked = "***"
            print(f"{args.get}: {masked}")
        except (ValueError, RuntimeError) as e:
            print(str(e), file=sys.stderr)
            sys.exit(1)
    elif args.check_1p:
        available = is_1password_available()
        print(f"1Password CLI: {'available' if available else 'not available'}")
        sys.exit(0 if available else 1)
    else:
        parser.print_help()
